test/ARCHITECTURE.md

```diff
@@ -0,0 +1,259 @@
+ directly but builds the required command for the `TEST_BROWSER` to start
+ either Chromium or Firefox. When a test calls any method on the browser object
+ the browser will be started, so tests which require no browser don't start a
+ browser needlessly.
+
+The `CDP` class is responsible for spawning the browser, then spawning a CDP
+driver this uses the `chrome-remote-interface` npm module to send commands to
+the spawned drivers via standard in (stdin).
+
+On `tearDown` the test status is inspected, if it failed test logs are
+collected and if the user has passed `--sit` the test pauses execution until
+the user presses enter so that the machine/browser state can be inspected. The
+test browser is killed after the `tearDown` function completed.
+
+Virtual machines are killed by the `TapRunner` once all tests have finished or
+in `setUp` if it's a `destructive` test as `nondestructive` tests re-use the
+existing global machine.
+
+### Test runner
+
+Cockpit uses a custom test runner to run the tests, spread the load over jobs
+and special handling of test failures. The test runner is implemented in
+Python in [test/common/run-tests](./common/run-tests) and expects a list of tests to be provided.
+
+The provided tests are collected and split up in `destructive` and
+`nondestrictive` tests and initialized as `Test` object. If there are any
+changes compared to the `main` branch, the test runner checks if any of the
+tests changed if so they are added to the affected test list unless more then
+three tests are changed. If `pkg/apps` is changed, `test/verify/check-apps`
+will also be added to the affected test list 
+
+After having collected the destructive, nondestructive and affected tests a
+scheduling loop is started, if a machine was provided it is used for the
+nondestructive tests, destructive tests will always spawn a new machine. If no
+machine is provided a pool of global machines is created based on the provided
+`--jobs` and nondestructive tests. The test runner will first try to assign all
+nondestructive tests on the available global machines and start the tests. 
+
+A test is started by the `Test` class by calling the `start()` method executed
+the provided `command` (e.g. `./test/verify/check-apps --machine 127.0.0.1:2201
+--browser 127.0.0.1:9091`) with a `timeout` to cancel hanging tests after a
+timeout automatically and creates a temporarily file to store the results of
+the test `command` in. Finally the test is added to the `running_tests` list.
+
+The test runner inspects all `running_tests` in a tight loop and polls if the
+test is still running. If the process stopped or exited the test output and
+`returncode` is saved. Depending on the `returncode` the test runner makes a
+decision on what to do next as described in the diagram below.
+```mermaid
+graph TD;
+    finished[Test finished]
+    succeeded{Test succeeded?}
+    affected{"Test affected?"}
+    affecteddone["Retry three times"]
+    skipped{Test skipped?}   
```

Comment:
This looks confusing. Normally "skipped" means to not run a test at all. It's also not naughty matching, as the test succeeded, and there's a separate "known issues" branch. According to the explanation below, the "skipped" branch should come first.